provider/balance_checker.go

```diff
@@ -2,153 +2,316 @@ package provider
+	btypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmrpc "github.com/tendermint/tendermint/rpc/core/types"
+
+	"github.com/ovrclk/akash/client"
 	"github.com/ovrclk/akash/provider/event"
 	"github.com/ovrclk/akash/provider/session"
 	"github.com/ovrclk/akash/pubsub"
 	"github.com/ovrclk/akash/util/runner"
-	"github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promauto"
-	"github.com/tendermint/tendermint/libs/log"
+	dtypes "github.com/ovrclk/akash/x/deployment/types/v1beta2"
+	"github.com/ovrclk/akash/x/escrow/client/util"
+	mtypes "github.com/ovrclk/akash/x/market/types/v1beta2"
 
-	"time"
+	aclient "github.com/ovrclk/akash/client"
+	netutil "github.com/ovrclk/akash/util/network"
 )
 
-var (
-	balanceGauge = promauto.NewGauge(prometheus.GaugeOpts{
-		Name: "provider_balance",
-	})
+type respState int
+
+const (
+	respStateNextCheck = iota
+	respStateOutOfFunds
+	respStateScheduledWithdraw
 )
 
+type BalanceCheckerConfig struct {
+	WithdrawalPeriod        time.Duration
+	LeaseFundsCheckInterval time.Duration
+}
+
+type leaseState struct {
+	tm                  *time.Timer
+	scheduledWithdrawAt time.Time
+}
+
 type balanceChecker struct {
-	session         session.Session
-	log             log.Logger
-	lc              lifecycle.Lifecycle
-	bus             pubsub.Bus
-	ownAddr         sdk.AccAddress
-	bankQueryClient bankTypes.QueryClient
-
-	cfg BalanceCheckerConfig
+	ctx     context.Context
+	session session.Session
+	log     log.Logger
+	lc      lifecycle.Lifecycle
+	bus     pubsub.Bus
+	ownAddr sdk.AccAddress
+	bqc     btypes.QueryClient
+	aqc     aclient.QueryClient
+	leases  map[mtypes.LeaseID]*leaseState
+	cfg     BalanceCheckerConfig
 }
 
-type BalanceCheckerConfig struct {
-	PollingPeriod           time.Duration
-	MinimumBalanceThreshold uint64
-	WithdrawalPeriod        time.Duration
+type leaseCheckResponse struct {
+	lid        mtypes.LeaseID
+	checkAfter time.Duration
+	state      respState
+	err        error
 }
 
 func newBalanceChecker(ctx context.Context,
-	bankQueryClient bankTypes.QueryClient,
+	bqc btypes.QueryClient,
+	aqc client.QueryClient,
 	accAddr sdk.AccAddress,
 	clientSession session.Session,
 	bus pubsub.Bus,
-	cfg BalanceCheckerConfig) *balanceChecker {
-	bc := &balanceChecker{
+	cfg BalanceCheckerConfig) (*balanceChecker, error) {
 
+	bc := &balanceChecker{
+		ctx:     ctx,
 		session: clientSession,
 		log:     clientSession.Log().With("cmp", "balance-checker"),
-		lc:      lifecycle.New(),
 		bus:     bus,
+		lc:      lifecycle.New(),
 		ownAddr: accAddr,
-
-		bankQueryClient: bankQueryClient,
-		cfg:             cfg,
+		bqc:     bqc,
+		aqc:     aqc,
+		leases:  make(map[mtypes.LeaseID]*leaseState),
+		cfg:     cfg,
 	}
 
+	startCh := make(chan error, 1)
 	go bc.lc.WatchContext(ctx)
-	go bc.run()
+	go bc.run(startCh)
+
+	select {
+	case err := <-startCh:
+		if err != nil {
+			return nil, err
```

Comment:
Can be `return bc, err` without the `if`. But I'm not sure you need this.

Reply:
to me it looks more convenient as `return bc, nil` clearly states at this point all is good.